### DISCUSSION TITLE
getting and setting cookies in POST/GET routes?

### DISCUSSION THREAD

**Discussion by simpel** [vercel/next.js#51889] [App Router]
HI all! 

How do I set cookies in my POST and GET routes? I figured something similar to the stuff below would work but it does not, whats the correct way of setting cookies in API endpoints? 

And also, when doing cookies().get("quizCookie") in a page.tsx it always returns undefined, so how to read a cookie in page.tsx?

```typescript
const defaultCookie: TQuizCookie = {
	currentStep: 0,
	correctAnswers: 0,
	totalSteps: 2,
}

export async function POST(request: NextRequest) {
	const response = NextResponse.next()

	response.cookies.set({
		name: 'quizCookie',
		value: JSON.stringify(defaultCookie),
		httpOnly: true,
		path: '/',
	})

	return response
}

export async function GET() {
	const response = NextResponse.next()
	response.cookies.get('quizCookie')
	return response
}

```

**Comment by icyJoseph** [comment]
First the actual code:

```ts
import { NextRequest, NextResponse } from "next/server";
import { cookies } from "next/headers";

const defaultCookie = {
  currentStep: 0,
  correctAnswers: 0,
  totalSteps: 2,
};

export async function POST(request: NextRequest) {
  const response = NextResponse.json({ ok: "ok" });

  response.cookies.set({
    name: "quizCookie",
    value: JSON.stringify(defaultCookie),
    httpOnly: true,
    path: "/",
  });

  return response;
}

export async function GET() {
  const cookie = cookies().get("quizCookie")?.value;
  console.log(cookie);
  const response = NextResponse.json({ ok: "ok", cookie });

  return response;
}
```

- Notice that `.next()` is avoided: https://nextjs.org/docs/messages/next-response-next-in-app-route-handler
- Notice, we use the cookies function to read the cookie: https://nextjs.org/docs/app/building-your-application/routing/router-handlers#cookies
- As to why it fails in page.tsx, I am not sure, perhaps, because it is never set? 
- Notice you also have to read the `.value` property.

**Reply by katiesun138** [reply]
Is it possible to use cookie/header to set a cookie in one API route and be able to use it in another API route? Is there a way of doing so? 
I generated a jwt token in my "login API route", then I return that token back to the page.tsx. I wanted to set it as a cookie on the client browser so that on every request (no matter which page.tsx), every page will be able to access the cookie value. I want to do this so that upon every GET/POST request, the cookie must be checked to see if it is a logged in user.

Thanks in advance!

**Reply by lucacavallaro** [reply]
@katiesun138 yes, it's possible. just use `cookies()` function as documented here https://nextjs.org/docs/app/building-your-application/routing/route-handlers#cookies

**Reply by devjiwonchoi** [reply]
@katiesun138 You can simply take @icyJoseph 's code and distribute it to different route handlers.
I'd suggest you use middleware though.

### Reference
- [middleware cookies](https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies)
